Smt/Tactic/Concretize.lean

```diff
@@ -0,0 +1,390 @@
+    there, instantiate the whole thing at the concrete args. pass these as list to tactic,
+    perhaps also heuristically choose hyps.
+- `concretizeApp`
+  - what about `Prop` args?
+  - what about mvars in args?
+-/
+def evalConcretize : ConcretizeM Unit := do
+  loop
+
+  if !(← read).rewriteWithLets then return
+
+  -- Sort and let-bind concretizations in dependency order.
+  let concs ← (← get).concretizations.foldM (init := #[]) fun acc nm conc =>
+    return acc.push (nm, conc, ← conc.getBody)
+  let concs := concs.qsort fun (_, conc₁, _) (_, _, eBody₂) => eBody₂.hasAnyFVar (· == conc₁.fvVar)
+
+  for (nm, conc, _) in concs do
+    -- Note: we must run this again here because the fvars may have changed.
+    let eBody ← Tactic.withMainContext conc.getBody
+    let fvLet ← liftMetaTacticAux fun mvarId => do
+      -- `$concNm := $eBody`
+      let (fv, mvarId) ← intro1P (← define mvarId nm (← inferType eBody) eBody)
+      return (fv, [mvarId])
+
+    withMainContext do
+      -- `var.$concNm = $concNm`
+      let eqTp ← mkEq (mkFVar conc.fvVar) (mkFVar fvLet)
+      let declEqBody ← conc.getDeclEqBody
+      let eq ← mkExpectedTypeHint declEqBody.toExpr eqTp
+      tryRewriteWhen eq fun ld => (`eqBody).isSuffixOf ld.userName.eraseMacroScopes
+
+    evalTactic <| ← `(tactic| clear $(mkIdent (nm ++ `eqBody)) $(mkIdent (nm ++ `eqVar)) $(mkIdent (nm ++ `var)))
+
+/-- Recursively finds and replaces concretizations in the goal and the hypotheses.
+We only concretize applications of the constants passed in as tactic arguments.
+
+TODO(WN): heuristics I tried for auto-choosing constants to concretize fail badly. What's a good one?
+
+A *concretization* of a function application `f #[a₁, .., aₙ]` is a new function or constant
+with the longest contiguous prefix of concretizable arguments to `f` fixed. An argument
+is *concretizable* when:
+- it has no free or bound variables (it's closed); and
+- it has forward dependencies, or it is a type argument.
+
+TODO(WN):
+- lift "contiguous" requirement?
+
+For example, `@List.foldl Nat Bool f` has the concretization
+  `List.foldl.Nat.bool : (Nat → Bool → Nat) → Nat → List Bool → Nat`
+and `@BitVec.xor 8 x y` has the concretization
+  `BitVec.xor.8 : BitVec 8 → BitVec 8 → BitVec 8`
+
+NB: In compilers this is *monomorphisation*, but besides types we also concretize dependent
+value arguments.
+
+The Coq variant is in `snipe`:
+- https://hal.archives-ouvertesfr/hal-03328935/document
+- https://github.com/smtcoq/sniper/blob/af7b0d22f496f8e7a0ee6ca495314d80ea2aa881/theories/elimination_polymorphism.v
+-/
+elab "concretize" "[" nms:ident,* "]" : tactic => do
+  let nms ← (nms : Array Syntax).mapM resolveGlobalConstNoOverloadWithInfo
+  let concretizeSet := nms.foldl (init := .empty) (NameSet.insert · ·)
+  evalConcretize
+    |>.run' { visitSet := #[.goal] }
+    |>.run { concretizeSet }
+
+end Smt.Concretize
+
+namespace test
```

Comment:
Is there a reason for putting the tests here and not in the `Test` directory? You can modify the current testing script if it is not sufficient...